### PR TITLE
ACMS-696: Post command hook approach to run rebuild after site install.

### PR DESCRIPTION
**Pull request by saurabhtripathics**
https://backlog.acquia.com/browse/ACMS-696 [SPIKE]
**Motivation**

_We currently rebuild cohesion assets after install in order to avoid timeout errors, but if we can run threaded tasks, we might improve our install experience.
 
_Issue_
 timeout errors

Fixes #NNN

**Proposed changes**
Investigate Symfony libraries for asynchronous threaded install tasks

-we could try drush post command hook and run cohesion rebuild from there
**Alternatives considered**
look in to Symfony\Component\Process\Process 

 Find if f there is a way to run a block of code directly (no we cannot call a class directly for example check docroot/core/tests/Drupal/Tests/Core/Command/QuickStartTest.php)

Another simpler option could be to modify the acquia_cms_rebuild_cohesion() function so that it tests if the request is coming from cli, and if so, just print a big terminal message telling the user they need to run drush cohesion:rebuild

**Testing steps**

- Run site install and rebuild should get run automatically after site insatll.

### CODE REVIEW

File: acquia_cms.profile [hunk 1]
```diff
@@ -46,7 +46,6 @@ function acquia_cms_form_cohesion_account_settings_form_alter(array &$form) {
     // configuration import and cohesion style rebuild functionality into
     // separate submit handlers.
     // @see \Drupal\cohesion_website_settings\Controller\WebsiteSettingsController::batch
-    $form['#submit'][] = 'acquia_cms_rebuild_cohesion';
```
Comment (gauravgoyal) [inline acquia_cms.profile:49]: @saurabhtripathics - why is this removed?

File: drush/Commands/SiteInstallCommands.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+class SiteInstallCommands extends DrushCommands {
+
+  /**
+   * Do a forceful rebuild whenever site is installed.
+   *
+   * @hook post-command site-install
+   */
+  public function postCommand($result, CommandData $commandData) {
+    $this->yell('Finished rebuilding.');
```
Comment (gauravgoyal) [inline drush/Commands/SiteInstallCommands.php:20]: @saurabhtripathics - let's also add some additional checks here please -

1. check if profile being installed is acquia_cms. We don't want this to trigger for any other profile install
2. check if site studio modules are already installed

**Review by gauravgoyal — CHANGES_REQUESTED** [review]

File: acquia_cms.profile (hunk 1 above)
Comment (saurabhtripathics) [reply acquia_cms.profile]: this code should be there, was experimenting something on local, added this code.

File: drush/Commands/SiteInstallCommands.php (hunk 2 above)
Comment (saurabhtripathics) [reply drush/Commands/SiteInstallCommands.php]: yes done these changes

File: drush/Commands/SiteInstallCommands.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+class SiteInstallCommands extends DrushCommands {
+
+  /**
+   * Do a forceful rebuild whenever site is installed.
+   *
+   * @hook post-command site-install
+   */
+  public function postCommand($result, CommandData $commandData) {
+    $profile = \Drupal::service('config.factory')->get('core.extension')->get('profile');
```
Comment (gauravgoyal) [inline drush/Commands/SiteInstallCommands.php:20]: @saurabhtripathics - you don't need to get the profile from here. It should be available as part of CommandData.

**Review by gauravgoyal — CHANGES_REQUESTED** [review]

File: drush/Commands/SiteInstallCommands.php (hunk 3 above)
Comment (saurabhtripathics) [reply drush/Commands/SiteInstallCommands.php]: done

File: drush/Commands/SiteInstallCommands.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * Do a forceful rebuild whenever site is installed.
+   *
+   * @hook post-command site-install
+   */
+  public function postCommand($result, CommandData $commandData) {
+    $arguments = $commandData->arguments();
+    $moduleHandler = \Drupal::service('module_handler');
+    if (isset ($arguments['profile'][0]) && $arguments['profile'][0] == 'acquia_cms' && $moduleHandler->moduleExists('cohesion')) {
```
Comment (gauravgoyal) [inline drush/Commands/SiteInstallCommands.php:22]: @saurabhtripathics - there shouldn't be a space after `isset` - should be `isset()`, right?

File: drush/Commands/SiteInstallCommands.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+use Consolidation\AnnotatedCommand\CommandResult;
+use Drupal\cohesion\Drush\DX8CommandHelpers;
+
+/**
+ * Rebuild site studio when site is installed via drush.
+ */
+class SiteInstallCommands extends DrushCommands {
+
+  /**
+   * Do a forceful rebuild whenever site is installed.
```
Comment (gauravgoyal) [inline drush/Commands/SiteInstallCommands.php:15]: @saurabhtripathics - comment could be - "Execute site studio rebuild after Acquia CMS site install."

**Review by gauravgoyal — CHANGES_REQUESTED** [review]

File: drush/Commands/SiteInstallCommands.php (hunk 4 above)
Comment (saurabhtripathics) [reply drush/Commands/SiteInstallCommands.php]: fixed

File: drush/Commands/SiteInstallCommands.php (hunk 5 above)
Comment (saurabhtripathics) [reply drush/Commands/SiteInstallCommands.php]: changed

**Comment by gauravgoyal** [comment]
@msherron @kdruckman  - can you take a look at this approach of rebuilding the site after site install?

File: drush/Commands/SiteInstallCommands.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+   */
+  public function postCommand($result, CommandData $commandData) {
+    $arguments = $commandData->arguments();
+    $moduleHandler = \Drupal::service('module_handler');
+    if (isset($arguments['profile'][0]) && $arguments['profile'][0] == 'acquia_cms' && $moduleHandler->moduleExists('cohesion')) {
+      // Forcefully clear the cache after site is installed otherwise site
+      // studio fails to rebuild.
+      drupal_flush_all_caches();
+      // Below code ensure that drush batch process doesn't hang. Unset all the
+      // ealier created batches so that drush_backend_batch_process() can run
```
Comment (kdruckman) [inline drush/Commands/SiteInstallCommands.php:27]: Minor nit: Typos. "ensures" and "earlier"

Comment (kdruckman) [reply drush/Commands/SiteInstallCommands.php:27]: Other than these minor typos, I think this approach makes sense to me. I tested it and it works. I think bringing this rebuild back into the site-install command will be much more straightforward for people used to that command.

**Review by kdruckman — CHANGES_REQUESTED** [review]

File: drush/Commands/SiteInstallCommands.php (hunk 6 above)
Comment (saurabhtripathics) [reply drush/Commands/SiteInstallCommands.php]: thanks, fixed typos

**Comment by saurabhtripathics** [comment]
All reviews done, this can be moved forward.

**Review by gauravgoyal — APPROVED** [review]

**Review by vishalkhode1 — APPROVED** [review]
